File: check-status.js
/**
 * Status checker
 *   1) Passport Seva application status
 *   2) GST ARN application status
 *
 * Usage: npm start
 *
 * GST captcha: script tries OCR; if unsure, type the characters shown in the browser into the terminal.
 */

require("dotenv").config({ quiet: true });
const readline = require("readline");
const { chromium } = require("playwright");
const Tesseract = require("tesseract.js");
const sharp = require("sharp");

const LOGIN_URL = "https://services1.passportindia.gov.in/forms/login";
const GST_URL = "https://services.gst.gov.in/services/arnstatus";

const USERNAME = process.env.PASSPORT_USER || "ABHISHEKSHPS@GMAIL.COM";
const PASSWORD = process.env.PASSPORT_PASS || "Mpasspw@01";
const FILE_NO = process.env.PASSPORT_FILE_NO || "LKN067803930926";
const GST_ARN = process.env.GST_ARN || "AA090726251099S";

function ask(question) {
  const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
  return new Promise((resolve) => {
    rl.question(question, (answer) => {
      rl.close();
      resolve((answer || "").trim());
    });
  });
}

function valueAfterLabel(bodyText, label) {
  const re = new RegExp(
    `${label}\\s*\\n\\s*-\\s*\\n\\s*([\\s\\S]*?)(?=\\n\\s*[A-Z][A-Za-z /()]+\\s*\\n\\s*-\\s*\\n|\\n\\s*Close\\b|\\n\\s*Home\\b|$)`,
    "i"
  );
  const m = bodyText.match(re);
  if (!m) return null;
  return m[1].replace(/\s+/g, " ").trim();
}

async function clickByExactText(page, text) {
  const pressable = page.locator('div[tabindex="0"]').filter({ hasText: new RegExp(`^${text}$`) });
  if (await pressable.count()) {
    await pressable.first().click({ force: true });
    return;
  }
  await page.getByText(text, { exact: true }).first().click({ force: true });
}

/* -------------------- Passport -------------------- */

async function loginPassport(page) {
  await page.goto(LOGIN_URL, { waitUntil: "networkidle", timeout: 60000 });
  await page.waitForTimeout(1000);

  const userInput = page.locator('input[type="text"]').first();
  const passInput = page.locator('input[type="password"]').first();
  await userInput.waitFor({ state: "visible", timeout: 30000 });

  await userInput.click();
  await userInput.fill("");
  await userInput.type(USERNAME, { delay: 15 });

  await passInput.click();
  await passInput.fill("");
  await passInput.type(PASSWORD, { delay: 15 });

  await clickByExactText(page, "Sign In");

  await page.waitForURL(/Home|homeScreen/i, { timeout: 60000 });
  await page.waitForFunction(
    () => /Submitted Applications|Welcome back/i.test(document.body.innerText),
    null,
    { timeout: 60000 }
  );
  await page.waitForTimeout(1500);
}

async function openPassportApplication(page) {
  await page.waitForFunction(
    (fileNo) => !fileNo || document.body.innerText.includes(fileNo),
    FILE_NO,
    { timeout: 30000 }
  );

  const viewButtons = page.locator('div[tabindex="0"]').filter({ hasText: /^View$/ });
  const count = await viewButtons.count();

  if (!count) {
    await page.getByText("View", { exact: true }).first().click({ force: true });
  } else if (FILE_NO) {
    let clicked = false;
    for (let i = 0; i < count; i++) {
      const btn = viewButtons.nth(i);
      const hasFile = await btn.evaluate((el, fileNo) => {
        let node = el;
        for (let d = 0; d < 12 && node; d++) {
          if ((node.innerText || "").includes(fileNo)) return true;
          node = node.parentElement;
        }
        return false;
      }, FILE_NO);
      if (hasFile) {
        await btn.click({ force: true });
        clicked = true;
        break;
      }
    }
    if (!clicked) await viewButtons.first().click({ force: true });
  } else {
    await viewButtons.first().click({ force: true });
  }

  await page.waitForURL(/UserApplicationsDetails/i, { timeout: 60000 });
  await page.waitForFunction(
    () => /Track Application Status/i.test(document.body.innerText),
    null,
    { timeout: 60000 }
  );
  await page.waitForTimeout(1000);
}

async function openPassportTrackStatus(page) {
  await clickByExactText(page, "Track Application Status");
  await page.waitForURL(/TrackApplication/i, { timeout: 60000 });
  await page.waitForFunction(
    () => /Status Tracker/i.test(document.body.innerText),
    null,
    { timeout: 60000 }
  );
  await page.waitForTimeout(800);
}

async function checkPassport(browser) {
  console.log("\n----- 1) PASSPORT STATUS -----");
  console.log("Opening Passport Seva login…");

  const context = await browser.newContext({
    viewport: { width: 1280, height: 900 },
    ignoreHTTPSErrors: true,
  });
  const page = await context.newPage();
  page.setDefaultTimeout(60000);
  page.on("dialog", async (d) => {
    console.log("Dialog:", d.message());
    await d.accept();
  });

  try {
    console.log("Signing in…");
    await loginPassport(page);
    console.log("Logged in. Opening application (View)…");
    await openPassportApplication(page);

    const detailsBody = await page.locator("body").innerText();
    const paymentStatus = valueAfterLabel(detailsBody, "Payment Status") || "(not found)";

    console.log("Opening Track Application Status…");
    await openPassportTrackStatus(page);
    const statusBody = await page.locator("body").innerText();
    const applicationStatus = valueAfterLabel(statusBody, "Status");
    const fileNumber = valueAfterLabel(statusBody, "File Number");

    if (!applicationStatus) throw new Error("Could not read Passport Status");

    return { fileNumber, paymentStatus, applicationStatus };
  } finally {
    await context.close();
  }
}

/* -------------------- GST -------------------- */

async function ocrGstCaptcha(pngBuffer) {
  try {
    const { data, info } = await sharp(pngBuffer).ensureAlpha().raw().toBuffer({ resolveWithObject: true });
    const out = Buffer.alloc(info.width * info.height);
    for (let i = 0, p = 0; i < data.length; i += info.channels, p++) {
      const r = data[i];
      const g = data[i + 1];
      const b = data[i + 2];
      if (r > 140 && g < 120 && b < 120) {
        out[p] = 255;
        continue;
      }
      const lum = 0.299 * r + 0.587 * g + 0.114 * b;
      out[p] = lum < 95 ? 0 : 255;
    }

    const cleaned = await sharp(out, {
      raw: { width: info.width, height: info.height, channels: 1 },
    })
      .resize({ width: info.width * 3, height: info.height * 3, kernel: "nearest" })
      .png()
      .toBuffer();

    const worker = await Tesseract.createWorker("eng");
    await worker.setParameters({
      tessedit_char_whitelist: "0123456789",
      tessedit_pageseg_mode: Tesseract.PSM.SINGLE_LINE,
    });
    const result = await worker.recognize(cleaned);
    await worker.terminate();

    const digits = (result.data.text || "").replace(/\D/g, "");
    if (digits.length >= 5 && digits.length <= 7) return digits;
    return "";
  } catch {
    return "";
  }
}

async function refreshGstCaptcha(page) {
  // Refresh icon is usually the 2nd button next to captcha image
  const refreshBtn = page.locator("button").filter({ has: page.locator("img, i, span, svg") });
  // Safer: reload captcha by clicking image's sibling refresh, or change img src
  const clicked = await page.evaluate(() => {
    const img = document.querySelector("#imgCaptcha");
    if (!img) return false;
    // Find nearby button with title/refresh
    const parent = img.parentElement;
    if (!parent) return false;
    const buttons = parent.querySelectorAll("button, a, i, span");
    for (const b of buttons) {
      const t = (b.getAttribute("title") || b.className || b.id || "").toLowerCase();
      if (t.includes("refresh") || t.includes("reload") || t.includes("captcha")) {
        b.click();
        return true;
      }
    }
    // Fallback: force new captcha URL
    img.src = "https://services.gst.gov.in/services/captcha?rnd=" + Math.random();
    return true;
  });
  if (!clicked) {
    await page.locator("#imgCaptcha").evaluate((img) => {
      img.src = "https://services.gst.gov.in/services/captcha?rnd=" + Math.random();
    });
  }
  await page.waitForTimeout(1000);
}

function parseGstField(bodyText, label) {
  // GST result rows look like: "Status\tPending for Processing"
  const re = new RegExp(`${label}\\s*[:：]?\\s*\\t\\s*([^\\n]+)`, "i");
  const m = bodyText.match(re) || bodyText.match(new RegExp(`${label}\\s*[:：]\\s*([^\\n]+)`, "i"));
  if (!m) return null;
  return m[1].replace(/\s+/g, " ").trim();
}

function parseGstStatus(bodyText) {
  const status = parseGstField(bodyText, "Status");
  if (status && !/^status$/i.test(status)) return status;
  if (/Pending for Processing/i.test(bodyText)) return "Pending for Processing";
  return null;
}

async function checkGst(browser) {
  console.log("\n----- 2) GST APPLICATION STATUS -----");
  console.log("Opening GST ARN status page…");

  const context = await browser.newContext({
    viewport: { width: 1280, height: 900 },
    ignoreHTTPSErrors: true,
  });
  const page = await context.newPage();
  page.setDefaultTimeout(60000);

  try {
    await page.goto(GST_URL, { waitUntil: "networkidle", timeout: 90000 });
    await page.waitForTimeout(1000);

    await page.locator("#ifARN").check({ force: true }).catch(() => {});
    await page.locator("#arnp").fill(GST_ARN);

    // Captcha appears after ARN is entered
    await page.locator("#imgCaptcha").waitFor({ state: "visible", timeout: 20000 });
    await page.locator("#fo-captcha").waitFor({ state: "visible", timeout: 20000 });

    const maxAttempts = 5;
    for (let attempt = 1; attempt <= maxAttempts; attempt++) {
      await page.waitForTimeout(500);
      const captchaBuf = await page.locator("#imgCaptcha").screenshot();
      let code = await ocrGstCaptcha(captchaBuf);

      if (code) {
        console.log(`OCR captcha guess (attempt ${attempt}): ${code}`);
        const override = await ask("Press Enter to use guess, OR type captcha from browser: ");
        if (override) code = override;
      } else {
        console.log(`Look at the captcha in the browser (attempt ${attempt}/${maxAttempts}).`);
        code = await ask("Type the captcha numbers here, then press Enter: ");
      }

      if (!code) {
        console.log("Empty captcha — refreshing…");
        await refreshGstCaptcha(page);
        continue;
      }

      await page.locator("#arnp").fill(GST_ARN);
      await page.locator("#fo-captcha").fill("");
      await page.locator("#fo-captcha").fill(code);
      await page.locator("#arnpreSearch").click();
      await page.waitForTimeout(3000);

      const body = await page.locator("body").innerText();

      if (/Enter valid Letters|Please enter the mandatory fields|Invalid Captcha|invalid captcha/i.test(body)
          && !/Pending for Processing|Approved|Rejected|Form No/i.test(body)) {
        console.log("Captcha rejected. Refreshing and retrying…");
        await refreshGstCaptcha(page);
        await page.locator("#fo-captcha").fill("");
        continue;
      }

      const status = parseGstStatus(body);
      if (!status) {
        throw new Error("GST search ran but Status was not found on the page");
      }

      return {
        arn: GST_ARN,
        status,
        formNo: parseGstField(body, "Form No\\.?"),
        formDesc: parseGstField(body, "Form Description"),
        submission: parseGstField(body, "Submission Date"),
      };
    }

    throw new Error("Failed to solve GST captcha after several attempts");
  } finally {
    await context.close();
  }
}

/* -------------------- Main -------------------- */

async function main() {
  const browser = await chromium.launch({ headless: false, slowMo: 40 });
  const results = { passport: null, gst: null };

  try {
    results.passport = await checkPassport(browser);
    results.gst = await checkGst(browser);

    console.log("\n========== FINAL STATUS ==========");
    console.log("1) PASSPORT");
    if (results.passport.fileNumber) {
      console.log("   File Number         :", results.passport.fileNumber);
    }
    console.log("   Payment Status     :", results.passport.paymentStatus);
    console.log("   Application Status :", results.passport.applicationStatus);
    console.log("");
    console.log("2) GST");
    console.log("   ARN                :", results.gst.arn);
    if (results.gst.formNo) console.log("   Form No.           :", results.gst.formNo);
    if (results.gst.formDesc) console.log("   Form Description   :", results.gst.formDesc);
    if (results.gst.submission) console.log("   Submission Date    :", results.gst.submission);
    console.log("   Application Status :", results.gst.status);
    console.log("==================================\n");
  } catch (err) {
    console.error("\nFailed:", err.message);
    process.exitCode = 1;
  } finally {
    await browser.close();
  }
}

main();

File: check_status.py
"""
Lightweight Passport / GST status checker (Render-friendly).

  python check_status.py --gst-only --headless --json
"""

from __future__ import annotations

import argparse
import gc
import hashlib
import io
import json
import os
import re
import sys
import time

import ddddocr
from dotenv import load_dotenv
from PIL import Image, ImageFilter, ImageOps
from playwright.sync_api import sync_playwright

load_dotenv()

LOGIN_URL = "https://services1.passportindia.gov.in/forms/login"
GST_URL = "https://services.gst.gov.in/services/arnstatus"

# Low-RAM Chromium flags for 512MB Render instances
CHROME_ARGS = [
    "--disable-dev-shm-usage",
    "--disable-gpu",
    "--no-sandbox",
    "--single-process",
    "--no-zygote",
    "--disable-extensions",
    "--disable-background-networking",
    "--disable-default-apps",
    "--disable-sync",
    "--disable-translate",
    "--mute-audio",
    "--no-first-run",
    "--renderer-process-limit=1",
    "--disable-features=TranslateUI,BlinkGenPropertyTrees,IsolateOrigins,site-per-process",
    "--js-flags=--max-old-space-size=96",
]


def env(name: str, default: str = "") -> str:
    return (os.getenv(name) or default).strip()


def digits_only(text: str) -> str:
    return re.sub(r"\D", "", text or "")


def alnum_only(text: str) -> str:
    return re.sub(r"[^A-Za-z0-9]", "", text or "")


def value_after_label(body: str, label: str) -> str | None:
    m = re.search(
        rf"{label}\s*\n\s*-\s*\n\s*([\s\S]*?)(?=\n\s*[A-Z][A-Za-z /()]+\s*\n\s*-\s*\n|\n\s*Close\b|\n\s*Home\b|$)",
        body,
        re.I,
    )
    if not m:
        return None
    return re.sub(r"\s+", " ", m.group(1)).strip()


def click_exact(page, text: str) -> None:
    pressable = page.locator('div[tabindex="0"]').filter(
        has_text=re.compile(rf"^{re.escape(text)}$")
    )
    if pressable.count():
        pressable.first.click(force=True)
    else:
        page.get_by_text(text, exact=True).first.click(force=True)


def launch_browser(playwright, headless: bool):
    return playwright.chromium.launch(
        headless=headless,
        args=CHROME_ARGS,
        chromium_sandbox=False,
    )


def check_passport(browser, username: str, password: str, file_no: str) -> dict:
    print("Passport: signing in…", flush=True)
    context = browser.new_context(
        viewport={"width": 800, "height": 600},
        ignore_https_errors=True,
    )
    page = context.new_page()
    page.set_default_timeout(45000)

    try:
        page.goto(LOGIN_URL, wait_until="domcontentloaded", timeout=60000)
        page.wait_for_selector('input[type="password"]', timeout=45000)
        # Prefer password login if OTP / Password tabs exist
        for label in ("Password", "Login with Password"):
            tab = page.get_by_text(label, exact=True)
            if tab.count():
                try:
                    tab.first.click(force=True)
                    time.sleep(0.3)
                except Exception:
                    pass
                break
        page.locator('input[type="text"]').first.fill(username)
        page.locator('input[type="password"]').first.fill(password)
        click_exact(page, "Sign In")
        page.wait_for_url(re.compile(r"Home|homeScreen", re.I), timeout=60000)
        page.wait_for_function(
            "() => /Submitted Applications|Welcome back/i.test(document.body.innerText)",
            timeout=45000,
        )

        print("Passport: opening application…", flush=True)
        views = page.locator('div[tabindex="0"]').filter(has_text=re.compile(r"^View$"))
        clicked = False
        if file_no and views.count():
            for i in range(views.count()):
                btn = views.nth(i)
                if btn.evaluate(
                    """(el, fileNo) => {
                        let n = el;
                        for (let d = 0; d < 12 && n; d++) {
                          if ((n.innerText || '').includes(fileNo)) return true;
                          n = n.parentElement;
                        }
                        return false;
                    }""",
                    file_no,
                ):
                    btn.click(force=True)
                    clicked = True
                    break
        if not clicked:
            if views.count():
                views.first.click(force=True)
            else:
                page.get_by_text("View", exact=True).first.click(force=True)

        page.wait_for_url(re.compile(r"UserApplicationsDetails", re.I), timeout=45000)
        details = page.locator("body").inner_text()
        payment = value_after_label(details, "Payment Status") or "(not found)"

        click_exact(page, "Track Application Status")
        page.wait_for_url(re.compile(r"TrackApplication", re.I), timeout=45000)
        body = page.locator("body").inner_text()
        status = value_after_label(body, "Status")
        if not status:
            raise RuntimeError("Could not read Passport Status")
        return {
            "file_number": value_after_label(body, "File Number"),
            "payment_status": payment,
            "application_status": status,
        }
    finally:
        context.close()


class CaptchaSolver:
    """One OCR model by default; beta lazily loaded on repeated failures."""

    def __init__(self) -> None:
        print("Loading OCR…", flush=True)
        self.ocr = ddddocr.DdddOcr(show_ad=False)
        self.ocr_beta = None  # lazy — heavy for 512MB Render

    def _ensure_beta(self) -> None:
        if self.ocr_beta is not None:
            return
        try:
            print("Loading OCR beta model…", flush=True)
            self.ocr_beta = ddddocr.DdddOcr(show_ad=False, beta=True)
        except Exception as exc:
            print(f"OCR beta unavailable: {exc}", flush=True)
            self.ocr_beta = False  # do not retry

    @staticmethod
    def _variants(png: bytes) -> list[bytes]:
        out: list[bytes] = [png]
        try:
            img = Image.open(io.BytesIO(png)).convert("RGB")
            w, h = img.size
            up = img.resize((w * 2, h * 2), Image.LANCZOS)
            gray = ImageOps.grayscale(up)
            threshed = gray.point(lambda p: 0 if p < 140 else 255).filter(
                ImageFilter.SHARPEN
            )
            for candidate in (up, gray, threshed):
                buf = io.BytesIO()
                candidate.convert("RGB").save(buf, format="PNG")
                out.append(buf.getvalue())
        except Exception:
            pass
        return out

    def guesses(self, png: bytes, use_beta: bool = False) -> list[str]:
        if use_beta:
            self._ensure_beta()
        seen: list[str] = []
        models = [self.ocr]
        if use_beta and self.ocr_beta and self.ocr_beta is not False:
            models.append(self.ocr_beta)
        variants = self._variants(png) if use_beta else [png]
        for variant in variants:
            for model in models:
                try:
                    raw = model.classification(variant) or ""
                except Exception:
                    raw = ""
                code = alnum_only(raw)[:8]
                if code and code not in seen:
                    seen.append(code)
        seen.sort(key=lambda g: (abs(len(g) - 6), -len(g)))
        return seen


GST_FIELD_BLACKLIST = {
    "-",
    "search",
    "back",
    "close",
    "home",
    "track application status",
    "track return status",
    "track payment status",
    "logout",
    "help",
}

KNOWN_GST_STATUSES = [
    "Pending for Processing",
    "Pending for Order",
    "Pending for Clarification",
    "Clarification filed - Pending for Order",
    "Approved",
    "Rejected",
    "Migrated",
    "Cancelled",
    "Provisional",
    "Withdrawn",
    "Validation Error",
]


def parse_gst_field(body: str, label: str) -> str | None:
    # Anchor label at line start so "Track Application Status" is never matched
    anchor = r"(?:^|\n)\s*"
    patterns = [
        rf"{anchor}{label}\s*[:：]?\s*\t+\s*([^\n\t]+)",
        rf"{anchor}{label}\s*[:：]\s*([^\n]+)",
        rf"{anchor}{label}\s*\n\s*([^\n]+)",
    ]
    for pat in patterns:
        for m in re.finditer(pat, body, re.I):
            value = re.sub(r"\s+", " ", m.group(1)).strip().strip("|:")
            if not value:
                continue
            low = value.lower()
            if low == label.lower() or low in GST_FIELD_BLACKLIST:
                continue
            return value
    return None


def detect_gst_status(body: str) -> str | None:
    for name in KNOWN_GST_STATUSES:
        if re.search(rf"\b{re.escape(name)}\b", body, re.I):
            return name
    return parse_gst_field(body, "Status")


def _captcha_hash(page) -> str:
    try:
        return hashlib.md5(page.locator("#imgCaptcha").screenshot()).hexdigest()
    except Exception:
        return ""


def refresh_gst_captcha(page, prev_hash: str = "") -> str:
    for _ in range(3):
        try:
            page.locator("button").filter(has=page.locator("i.fa-refresh")).first.click(
                force=True
            )
        except Exception:
            try:
                page.locator("#imgCaptcha").click(force=True)
            except Exception:
                pass
        # Wait for a NEW captcha image (different bytes)
        for _ in range(20):
            time.sleep(0.15)
            h = _captcha_hash(page)
            if h and h != prev_hash:
                return h
        prev_hash = _captcha_hash(page)
    return prev_hash


def check_gst(browser, solver: CaptchaSolver, arn: str) -> dict:
    print("GST: opening ARN page…", flush=True)
    context = browser.new_context(
        viewport={"width": 800, "height": 600},
        ignore_https_errors=True,
        java_script_enabled=True,
    )
    # Block heavy third-party fonts / trackers to save memory & time
    def _route(route):
        try:
            r = route.request
            url = r.url
            rtype = r.resource_type
            if rtype in ("font", "media"):
                return route.abort()
            if "google-analytics" in url or "googletagmanager" in url or "gstatic" in url:
                return route.abort()
            return route.continue_()
        except Exception:
            try:
                route.continue_()
            except Exception:
                pass

    context.route("**/*", _route)
    page = context.new_page()
    page.set_default_timeout(45000)

    try:
        page.goto(GST_URL, wait_until="domcontentloaded", timeout=60000)
        page.locator("#ifARN").check(force=True)
        page.locator("#arnp").fill(arn)
        page.locator("#imgCaptcha").wait_for(state="visible", timeout=20000)
        time.sleep(0.6)

        max_attempts = 20
        current_hash = _captcha_hash(page)

        for attempt in range(1, max_attempts + 1):
            png = page.locator("#imgCaptcha").screenshot()
            guesses = solver.guesses(png, use_beta=attempt >= 4)
            code = guesses[0] if guesses else ""
            print(f"GST captcha try {attempt}: {code!r} (from {guesses})", flush=True)

            if not code or len(code) < 4:
                current_hash = refresh_gst_captcha(page, current_hash)
                continue

            page.locator("#arnp").fill(arn)
            page.locator("#fo-captcha").fill(code)
            page.locator("#arnpreSearch").click()
            time.sleep(2.5)
            body = page.locator("body").inner_text()

            status = detect_gst_status(body)
            form_no = parse_gst_field(body, r"Form No\.?")
            form_desc = parse_gst_field(body, "Form Description")
            submission = parse_gst_field(
                body, r"(?:Date of Submission|Submission Date)"
            )
            has_known_status = any(
                s.lower() in body.lower() for s in KNOWN_GST_STATUSES
            )
            result_visible = (
                "Search Result based on ARN" in body
                or has_known_status
                or (status and (form_no or submission))
            )
            print(
                f"GST parsed: status={status!r} form_no={form_no!r} "
                f"submission={submission!r} known={has_known_status}",
                flush=True,
            )
            if result_visible and status:
                return {
                    "arn": arn,
                    "status": status,
                    "form_no": form_no,
                    "form_desc": form_desc,
                    "submission": submission,
                }

            current_hash = refresh_gst_captcha(page, current_hash)

        raise RuntimeError("Could not solve GST captcha")
    finally:
        context.close()


def format_status_message(passport: dict | None, gst: dict | None) -> str:
    lines = ["========== STATUS =========="]
    if passport:
        lines.append("PASSPORT")
        if passport.get("error"):
            lines.append(f"  Error: {passport['error']}")
        else:
            lines += [
                f"  File: {passport.get('file_number') or '-'}",
                f"  Payment: {passport.get('payment_status')}",
                f"  Status: {passport.get('application_status')}",
            ]
    if gst:
        if passport:
            lines.append("")
        lines.append("GST")
        if gst.get("error"):
            lines += [f"  ARN: {gst.get('arn') or '-'}", f"  Error: {gst['error']}"]
        else:
            lines += [
                f"  ARN: {gst.get('arn')}",
                f"  Form: {gst.get('form_no') or '-'}",
                f"  Submitted: {gst.get('submission') or '-'}",
                f"  Status: {gst.get('status')}",
            ]
    lines.append("============================")
    return "\n".join(lines)


def run_status_check(headless: bool = False, gst_only: bool = True) -> dict:
    username = env("PASSPORT_USER")
    password = env("PASSPORT_PASS")
    file_no = env("PASSPORT_FILE_NO")
    arn = env("GST_ARN")

    if not arn:
        raise RuntimeError("Missing GST_ARN")

    solver = CaptchaSolver()
    passport = None
    gst = None
    gst_error = None

    with sync_playwright() as p:
        # One browser at a time — frees RAM between steps
        if not gst_only:
            if not all([username, password, file_no]):
                passport = {"error": "Missing passport credentials"}
            else:
                browser = launch_browser(p, headless)
                try:
                    passport = check_passport(browser, username, password, file_no)
                except Exception as exc:
                    passport = {"error": str(exc) or "Passport check failed"}
                    print(f"Passport failed: {passport['error']}", flush=True)
                finally:
                    browser.close()
                    gc.collect()

        browser = launch_browser(p, headless)
        try:
            gst = check_gst(browser, solver, arn)
        except Exception as exc:
            gst_error = str(exc) or "GST check failed"
            print(f"GST failed: {gst_error}", flush=True)
        finally:
            browser.close()
            gc.collect()

    if gst is None and passport is None:
        raise RuntimeError(gst_error or "No status returned")
    if gst is None:
        gst = {"arn": arn, "error": gst_error or "GST check failed"}
    return {"passport": passport, "gst": gst}


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--headless", action="store_true")
    parser.add_argument("--json", action="store_true")
    parser.add_argument("--gst-only", action="store_true", help="Only check GST (default for API)")
    parser.add_argument("--with-passport", action="store_true", help="Also check passport")
    args = parser.parse_args()
    # Default: GST only. Passport runs ONLY with explicit --with-passport.
    gst_only = not args.with_passport
    print(f"Mode: {'GST only' if gst_only else 'Passport + GST'}", flush=True)

    if args.json:
        sys.stdout = sys.stderr

    try:
        result = run_status_check(headless=args.headless, gst_only=gst_only)
        message = format_status_message(result.get("passport"), result.get("gst"))
        if args.json:
            sys.stdout = sys.__stdout__
            print(json.dumps({"ok": True, "message": message, **result}))
        else:
            print("\n" + message + "\n")
    except Exception as exc:
        if args.json:
            sys.stdout = sys.__stdout__
            print(json.dumps({"ok": False, "error": str(exc)}))
            sys.exit(1)
        raise


if __name__ == "__main__":
    main()
